### PR TITLE
PS-7657 crash seen with update query on a partition table having comp…

### PR DESCRIPTION
**Pull request by nitendra-bhosle**
…ressed column

PS-7657 crash seen with update query on a partition table having compressed column
    
    https://jira.percona.com/browse/PS-7657
    
    Analysis :
    Update statements execution flow on tables with partitions by key and having compressed column:
    
    1.First it searches for the row and positions the cursor (the blob is decompressed and stored in blob_heap and record is returned to server)
    2. Builds new row and then calls update(old_row, new_row). Since the blob field doesn't change, the blob from old row is stored in the new row (mysql record)
    3. Parition hanlder detects that the updated row should be in different partition
    4. Insert the new row in new parition. During insertion, we empty the blob heap. This releases the blocks used by blob and corrupts the blob data in new row
    5. So we end up inserting garbage to the new partition after the update
    
    Fix :
    Avoided prebuilt->compress_heap from being empty
    till it's required

### CODE REVIEW

File: storage/innobase/row/row0mysql.cc
Comment (satya-bodapati) [inline storage/innobase/row/row0mysql.cc:2304]: @nitendra-bhosle Can you please explain how the freed compress_heap causes update to fail (not find the record?)?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/row/row0mysql.cc
Comment (nitendra-bhosle) [reply storage/innobase/row/row0mysql.cc:2304]: @satya-bodapati 'mem_heap_empty' function Empties a memory heap but first memory block of the heap is not freed 
hence inside 'function row_compress_column' prebuilt->compress_heap is not NULL 
But memory is emptied.

Comment (satya-bodapati) [reply storage/innobase/row/row0mysql.cc:2304]: @nitendra-bhosle what happens after that? is the buffer insufficient to compress the column?

Comment (satya-bodapati) [reply storage/innobase/row/row0mysql.cc:2304]: and the question how this translates to update not being able to find the row still remains

Comment (nitendra-bhosle) [reply storage/innobase/row/row0mysql.cc:2304]: Inside method : 'row_compress_column'
buf = static_cast<byte*>(mem_heap_zalloc(
			prebuilt->compress_heap,buf_len));
this call operates on prebuilt->compress , which tries to memset memory , which it doesn't have ownership,
Which leads to memory pointed by pointer "const byte*	data" to be getting corrupted/emptied.
And hence buf returned from this method is empty .Which is causing update not being able to find the row

Comment (satya-bodapati) [reply storage/innobase/row/row0mysql.cc:2304]: how did you prove that it doesn't have ownership? 
memset() is working on new block of memory returned by heap_zalloc().

I am not sure if I understand the fix. row_mysql_prebuilt_free_compress_heap() seems to be called too.

Comment (nitendra-bhosle) [reply storage/innobase/row/row0mysql.cc:2304]: Inside function 'row_compress_column' 
buf = static_cast<byte*>(mem_heap_zalloc(
			prebuilt->compress_heap,buf_len));
this statement empties 'data' . Which happens when prebuilt->compress_heap doesn't have memory allocated.
I checked this by manually executing 
prebuilt->compress_heap =
			mem_heap_create(ut_max(UNIV_PAGE_SIZE, buf_len)); 
statement everythime.

'mem_heap_create'  create heap of size maximum of UNIV_PAGE_SIZE or buf_len , which is most likely to be UNIV_PAGE_SIZE

'mem_heap_zalloc' allocates and memsets  'buf_len' bytes out of heap block created by  'mem_heap_create'

'row_mysql_prebuilt_free_compress_heap' free's heap created by 'mem_heap_create' , this executes on end of statement.

Ideally there should be pair of 'mem_heap_create' and 'row_mysql_prebuilt_free_compress_heap' for each sql statement

Comment (satya-bodapati) [reply storage/innobase/row/row0mysql.cc:2304]: row_mysql_prebuilt_free_compress_heap() is already called at end of statement. once heap is created, other iterations need not bother about creating the heap again. mem_heap_t is dynamic memory allocator. ie it will allocate new blocks requested. Something else is going @nitendra-bhosle

Comment (satya-bodapati) [reply storage/innobase/row/row0mysql.cc:2304]: Specially since this happening with partitions, check if we need separate compressed heaps for paritions. 
Check m_blob_heap_parts/ clear_blob_heaps(). Do we need something like that?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/row/row0mysql.cc
Comment (nitendra-bhosle) [reply storage/innobase/row/row0mysql.cc:2304]: I will explore this.

Comment (satya-bodapati) [reply storage/innobase/row/row0mysql.cc:2304]: @nitendra-bhosle 

I debugged a bit and I think your fix makes sense but here is the explanation

So the way blobs work is storage engines only return a pointer to the blob (not make a copy) to the server. This pointer is from the prebuilt->blob_heap. If blob_heap is emptied, the blob returned by SE is corrupted.

Coming to the UPDATE statement:

1. First it searches for the row and positions the cursor (the blob is decompressed and stored in blob_heap and record is returned to server)
2. Builds new row and then calls update(old_row, new_row). Since the blob field doesn't change, the blob from old row is stored in the new row (mysql record)
3. Parition hanlder detects that the updated row should be in different partition
4. Insert the new row in new parition. During insertion, we empty the blob heap. This releases the blocks used by blob and corrupts the blob data in new row
5. So we end up inserting garbage to the new partition after the update

Requests
--------
1. Update commit message explaining the sequence clearly

2. Fix this blob heap code in row_sel_store_mysql_rec()
           if (UNIV_LIKELY_NULL(prebuilt->blob_heap)) {
                row_mysql_prebuilt_free_blob_heap(prebuilt);
        }

        if (UNIV_LIKELY_NULL(prebuilt->compress_heap))
                mem_heap_empty(prebuilt->compress_heap);
  
         call free for compress_heap as well
        
3. Bring the replication testcase from PS-7599

Comment (satya-bodapati) [reply storage/innobase/row/row0mysql.cc:2304]: 4. Run the mtr testcase on valgrind/ASAN and paste the result here.

Comment (nitendra-bhosle) [reply storage/innobase/row/row0mysql.cc:2304]:
```
mtr test-case on valgrind results 
./mtr --valgrind percona_bug_ps7599.test
==============================================================================

TEST                                      RESULT   TIME (ms) or COMMENT
--------------------------------------------------------------------------

worker[1] Using MTR_BUILD_THREAD 300, with reserved ports 13000..13009
[ 20%] rpl.percona_bug_ps7599 'mix'             [ skipped ]  Doesn't support --binlog-format='mixed'
[ 40%] rpl.percona_bug_ps7599 'row'             [ pass ]  15462
[ 60%] rpl.percona_bug_ps7599 'stmt'            [ skipped ]  Doesn't support --binlog-format='statement'
[ 80%] shutdown_report                          [ pass ]       
[100%] valgrind_report                          [ pass ]       
--------------------------------------------------------------------------
The servers were restarted 0 times
Spent 15.462 of 48 seconds executing testcases

Completed: All 3 tests were successful.

2 tests were skipped, 0 by the test itself.

./mtr --valgrind xtradb_compressed_columns.test
==============================================================================

TEST                                      RESULT   TIME (ms) or COMMENT
--------------------------------------------------------------------------

worker[1] Using MTR_BUILD_THREAD 300, with reserved ports 13000..13009
[ 33%] innodb.xtradb_compressed_columns         [ pass ]  54374
[ 66%] shutdown_report                          [ pass ]       
[100%] valgrind_report                          [ pass ]       
--------------------------------------------------------------------------
The servers were restarted 0 times
Spent 54.374 of 78 seconds executing testcases

Completed: All 3 tests were successful.
```

File: mysql-test/suite/rpl/t/percona_bug_ps7599.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/percona_bug_ps7599.test]: IMO, using JIRA number in the testcase name is not very intuitive and will make the person to actually look into the test while categorizing the failure in case the test fails in future.

Can we rename the testcase to `rpl_update_partitioned_table_with_compressed_column.test` or something like that?

File: mysql-test/suite/rpl/t/rpl_update_partitioned_table_with_compressed_column.test
Comment (satya-bodapati) [inline mysql-test/suite/rpl/t/rpl_update_partitioned_table_with_compressed_column.test]: remove this

File: mysql-test/suite/rpl/t/rpl_update_partitioned_table_with_compressed_column.test
Comment (satya-bodapati) [inline mysql-test/suite/rpl/t/rpl_update_partitioned_table_with_compressed_column.test]: just use name t1

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test
Comment (satya-bodapati) [inline mysql-test/suite/innodb/t/xtradb_compressed_columns.test]: use t1 as name

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test
Comment (satya-bodapati) [inline mysql-test/suite/innodb/t/xtradb_compressed_columns.test:233]: also add SELECT to ensure the blob stored isn't corrupted

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
LGTM! minor suggestions

File: mysql-test/suite/rpl/t/rpl_update_partitioned_table_with_compressed_column.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_update_partitioned_table_with_compressed_column.test]: Including`include/sync_slave_sql_with_master.inc` changes the connection to slave in the end.
So, this DELETE is executed on slave and not on the master.

So, please use
```
--source include/rpl_connection.inc
DROP TABLE tt_2_p;
--source include/rpl_end.inc
```

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/innodb/t/xtradb_compressed_columns.test]: Remove these blank lines in the end of the test.

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test
Comment (venkatesh-prasad-v) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test:233]: It is probably better to use asserts in addition to SELECTs.

File: mysql-test/suite/rpl/t/rpl_update_partitioned_table_with_compressed_column.test
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_update_partitioned_table_with_compressed_column.test]: In between these, can we validate that both master and slave have same data?

We can just source diff_tables.inc
```
--let $diff_tables= master:t1,slave:t1
--source include/diff_tables.inc
```

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test
Comment (satya-bodapati) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test:233]: either way is fine. at least for SELECTs. Otherwise, entire mtr suite should have been using asserts and never SELECTs. isn't it?

Abort on error is good for debugging, not so much from a coverage perspective.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
1. Please fix GitHub description
2. Please add Jenkins link as a comment

**Comment by nitendra-bhosle** [comment]
Jenkins : https://ps57.cd.percona.com/view/5.7/job/percona-server-5.7-param/446/

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test
Comment (george-lorch) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test:233]: Right, newer semantics, especially with built in mtr assert means that an explicit assert on single valued expressions is clearer in the test case and can not be washed away with a blind re-record.  So assert is preferred over recorded results for single value checks now.

Comment (satya-bodapati) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test:233]: if you want to cheat, add one more row to table :smiling_imp:

Comment (nitendra-bhosle) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test:233]: :) , I have added assert in test case

**Review by venkatesh-prasad-v — APPROVED** [review]
Thank you for addressing the review comments. The patch looks good and is approved from my side. But please wait for @satya-bodapati for the final approval.

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test
Comment (satya-bodapati) [inline mysql-test/suite/innodb/t/xtradb_compressed_columns.test:234]: count(*) doesn't verify contents. the garbage inserted can be still counted as 1. So please use content comparison instead of COUNT(*)

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: mysql-test/suite/innodb/t/xtradb_compressed_columns.test
Comment (nitendra-bhosle) [reply mysql-test/suite/innodb/t/xtradb_compressed_columns.test:234]: Inside Where condition I have used actual content , so count(*) will return 0 if content is not same as what we have inserted earlier.

**Review by satya-bodapati — APPROVED** [review]

**Review by satya-bodapati — APPROVED** [review]

**Pull request merged**
